File: common/src/main/java/io/github/moulberry/moulconfig/xml/loaders/TextFieldLoader.java
package io.github.moulberry.moulconfig.xml.loaders;

import io.github.moulberry.moulconfig.gui.component.TextFieldComponent;
import io.github.moulberry.moulconfig.xml.XMLContext;
import io.github.moulberry.moulconfig.xml.XMLGuiLoader;
import io.github.moulberry.moulconfig.xml.XMLUniverse;
import org.w3c.dom.Element;

import javax.xml.namespace.QName;

public class TextFieldLoader implements XMLGuiLoader<TextFieldComponent> {
    @Override
    public TextFieldComponent createInstance(XMLContext<?> context, Element element) {
        return new TextFieldComponent(
                context.getPropertyFromAttribute(element, new QName("value"), String.class),
                context.getPropertyFromAttribute(element, new QName("editable"), boolean.class),
                context.getPropertyFromAttribute(element, new QName("suggestion"), String.class, ""),
                context.getPropertyFromAttribute(element, new QName("width"), int.class, 80)
        );
    }

    @Override
    public QName getName() {
        return XMLUniverse.qName("TextField");
    }
}

File: common/src/main/java/io/github/moulberry/moulconfig/xml/loaders/TextFieldLoader.kt
package io.github.moulberry.moulconfig.xml.loaders

import io.github.moulberry.moulconfig.gui.component.TextFieldComponent
import io.github.moulberry.moulconfig.observer.GetSetter
import io.github.moulberry.moulconfig.xml.XMLContext
import io.github.moulberry.moulconfig.xml.XMLGuiLoader
import io.github.moulberry.moulconfig.xml.XMLUniverse
import org.w3c.dom.Element
import javax.xml.namespace.QName

class TextFieldLoader : XMLGuiLoader<TextFieldComponent?> {
    override fun createInstance(context: XMLContext<*>, element: Element): TextFieldComponent {
        return TextFieldComponent(
            context.getPropertyFromAttribute(element, QName("value"), String::class.java)!!,
            context.getPropertyFromAttribute(element, QName("width"), Int::class.javaPrimitiveType, 80),
            context.getPropertyFromAttribute(element, QName("editable"), Boolean::class.java)
                ?: GetSetter.constant(true),
            context.getPropertyFromAttribute(element, QName("suggestion"), String::class.java, ""),
        )
    }

    override fun getName(): QName {
        return XMLUniverse.qName("TextField")
    }
}
